### DISCUSSION TITLE
Considering the `if (!initialized)` pattern

### DISCUSSION THREAD

**Discussion by 414owen** [c3lang/c3c#1367] [General]
A lot of operations are defined on "uninitialized" types. For example `rand` checks whether the underlying structure is initialized, and if not, it seeds it, and sets the `initialized` flag.

This seems like unnecessary overhead and complexity to me. Firstly, the check itself. While it'll (probably) be branch-predicted correctly from the third use onwards, and might only add minimal overhead in microbenchmarks, branch predictor saturation exists, and in larger codebases, this will manifest as latency that a user won't track down.

Some types don't have an existing field which indicates whether they are initialized, meaning we have to store extra state, as is the case with the `initialized` boolean.

---

Proposed solution:

Allow the user to specify that the initialization of a type is mandatory, via an attribute on the type declaration.
This would (probably) mean having initializers that return a new value (eg. `new :: uint -> HashMap`), rather than initializing an uninitialized value.
Optionally, allow users to specify a default initializer, which would be used instead of the default zero-initialization.
Migrate the stdlib to this pattern, and remove lots of checks for `!initialized`.

**Comment by lerno** [comment]
In the case of "rand" it's special, since that is a convenience method for getting random numbers from a common thread local random, it's not an optimized function. For that you should instead use a custom random generator.

**Reply by 414owen** [reply]
Okay, thanks for the quick reply.
How about the checks (`if (!map.allocator)` and `if (!map.count)`) in `map.c3`?

Maybe an alternative approach would be  `@require !self.is_initialized()` annotations?
Although I guess those end up being runtime checks, rather than compile-time?

**Reply by lerno** [reply]
The checks in map is a convenience, allow it to be used without an explicit init. Whether this is good or not... Well, a larger question is really whether stdlib types in general should be provided by pointer or by value.

If the interface would be `map::new_map(...)` and you got a pointer, then this would be the initialization point quite naturally.

However, this would preclude value semantics *in the cases that is desired*. So should maps be by pointer? Maybe they should.

**Reply by 414owen** [reply]
Why does requiring an explicit init mean we would have to store the map by reference?

I would prefer to keep the outer layer of the HashMap on the stack, but I still want the compiler to force me to initialize it before use, so that HashMap methods don't have to check whether it has been initialized...

**Reply by lerno** [reply]
This is a convenient pattern:
```c
SomeMap x;
x["foo"] = "bar";
...
x.free();
do_something_using_map(abc, SomeMap {});
```
This is a bigger question than just this init, but is about what the *standard library* convenience types should do. Obviously the *most optimized* solution will rarely be the standard library one. So the question is how to make the standard library variants sufficiently easy to work with, in particular not mistakenly use them incorrectly.

I think you dropped into the discord? Would you like to participate in the discussion in `#standardlib`?

**Reply by 414owen** [reply]
I'm happy to chime in on discord, but I actually prefer a GitHub as a forum for something like this
Discord has people discussing different things at once, with overlap, and you have to piece together a thread...
 
To answer your point, yes, tracking freeing a value would also be necessary to enforce correct usage AND omit the init checks from some operations. Doing this correctly will involve the compiler tracking resource acquisition, moves, and releases. Whether it's possible to make a substructural type system ergonomic to use.... I really don't know.

**Reply by lerno** [reply]
I've already discussed this on Discord, but to recap, there are basically 4 designs possible for a container:
```c
// A - stack allocatable struct with internal data allocated (typically on the heap)
struct IntList { usz size, usz capacity, int* data }
IntList a;
a.init(16); // `a` is passed by ref, allows setting the allocator
a.append(1); // `a` is passed by ref
a.free(); // Frees anything in 'data'
// Possibly allowed
IntList b;
b.append(1); // Implicit init to use the heap, this would require "is initialized" checks.

// B - struct ONLY accessed through pointer, use "new", null is invalid
struct IntListImpl { usz size, usz capacity, int* data }
distinct IntList = IntListImpl*;
IntList a; // null
a = list::new(<int>)(16); // Allocates struct and 'data'
a.append(1); // a by value
a.free(); // frees struct and data

// C - struct ONLY accessed through pointer, but null is valid
struct IntListImpl { usz size, usz capacity, int* data }
distinct IntList = IntListImpl*;
IntList a; // null
a = list::new(<int>)(16); // Allocates struct and 'data'
a.append(1); // 'a' by ref
a.free(); // frees struct and data
IntList b = null;
b.append(1); // Implicit new

// D (stb style)
struct IntListHeader { usz size, usz capacity, int[] data }
distinct IntList = IntListHeader*;
IntList a;
a.init(16); // allocates and sets the pointer
// alternatively a = list::new(<int>)(16);
a.append(1); // 'a' by ref, may change even non-null pointers
// Possibly allowed
IntList b = null;
b.append(1);
```

**Reply by lerno** [reply]
All of these have slightly different behaviour and need for understanding them properly. 

(A) has the virtue of being stack declarable. (Although B-D can use stack and temp allocators easily, preventing need for heap allocation), however it needs to be explicitly passed by reference, and has by default a bigger memory footprint when embedded by value. I think C programmers will naturally gravitate towards this type.

(B) Is the "easy" type, which programmers from C++ or Java will feel comfortable with. It's the variant that needs the least amount of understanding the inner workings, and is the safest to use overall. It can use `free` to deallocate both its contained data and itself, which is nice. However, it uses a second level of indirection to use its data.

(C) is a variant of (B) but will mutate the pointer on append for null values, which means it's less safe than (B), although it has some conveniences, not the least that it works well with ZII.

(D) has similar requirements as C, but removes the need to indirection. It does this by being even less stable. Where C only mutates the underlying pointer for `null` values, D will mutate the underlying pointer at any time. This is a drawback it shares with (A).

All of the above might be the optimal solution *in some cases*, the question is what the standard library should provide by default.

**Reply by 414owen** [reply]
I would by far favour A, where the struct lives on the stack. I don't think the stdlib should be subject to premature pessimization.

Unless I'm misunderstanding, none of these options change the fact that you need a call to init the value, so none of them change the question of whether to do init checks in other methods or not.

Currently, an easy solution would be to say that the user has to call `a.init()`, and calling any other operations before doing so is undefined behaviour. Maybe at a later point it could become a compile-time error, to be a little friendlier.

**Reply by lerno** [reply]
Using (A) was my intuition, but I am questioning it. It's the hardest to get right, and there are other issues with it as well.

**Reply by lerno** [reply]
There is also the enhanced ABI stability of using B, C or D.

**Reply by 414owen** [reply]
> There is also the enhanced ABI stability of using B, C or D.

Won't a change in the structure result in ABI breakage whether the outer layer is on the heap or not? Where does the extra stability come from?

**Reply by lerno** [reply]
If the type is opaque, then all manipulation of the data occurs through functions, which are provided by the library itself. 

Imagine you compile for 1.0 and dynamically link with it. You create with `new` do an `append` which calls the `append` function. This calls 1.0 new, giving you a pointer to a 1.0 object and then invokes 1.0 append.

Now this program is run with 1.1. This means it now calls 1.1 new, will get the 1.1 data, and 1.1 append.

**Reply by 414owen** [reply]
Okay I see what you mean. I personally wouldn't care much about being able to use different dynamically linked versions of the stdlib without recompiling my code, so I wouldn't pick that over more performance. I can see the appeal though.
